ARTEMIS-1663 - Add new message count and size metrics

cshannon opened the pull request:
Adding new metrics for tracking message counts and sizes on a Queue.
This includes tracking metrics for pending, delivering and scheduled
messages.  The paging store also tracks message size now.

cshannon commented:
This PR is quite large so I would appreciate if multiple people take a look to make sure it looks ok.

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageSubscriptionCounterImpl.java:
```diff
@@ -374,8 +403,43 @@ private ItemOper(PageSubscriptionCounterImpl counter, long id, int add) {
+
+   private static class PendingCounter {
+      final long id;
+      final AtomicInteger count;
```
could this be an int, with a static atomic updater (ditto with the atomic below also)

cshannon replied:
What do you mean by atomic updater, do you mean using something like AtomicIntegerFieldUpdater ?

cshannon replied:
I'm about to push an updated PR using the field updater classes.

michaelandrepearce replied:
Yes this i what i meant , ill have a look at the updated PR in a bit/tomorrow.

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageSubscriptionImpl.java:
```diff
@@ -1309,4 +1336,68 @@ public void remove() {
    }
+
+
+   private static class PagePositionSize {
```
Why not just extend PagePosition (saves on objects)

cshannon replied:
The reason I didn't extend is because PagePosition has methods like nextMessage() which increment the message number, etc.  If that is called then the stored size doesn't really make sense anymore as it won't match which is why I just created a new object.

michaelandrepearce replied:
But this new object still holds PagePosition so its not been removed.

cshannon replied:
I will fix this and extend PagePosition as if the position is incremented that is fine as the size will just go back to 0 unless set.

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/codec/PageCountRecordIncV2.java:
```diff
@@ -0,0 +1,72 @@
+import org.apache.activemq.artemis.api.core.ActiveMQBuffer;
+import org.apache.activemq.artemis.utils.DataConstants;
+
+public class PageCountRecordIncV2 extends PageCountRecordInc {
```
Is a new V2 needed? could have added size to PageCountRecord default to , and checked for readable bytes before read the long, to be back compatible.

cshannon replied:
Also a good point, I can get rid of the extra objects and just do a check on the buffer for readableBytes()  is greater than 0. Thanks.

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/codec/PageCountRecordV2.java:
```diff
@@ -0,0 +1,63 @@
+import org.apache.activemq.artemis.api.core.ActiveMQBuffer;
+import org.apache.activemq.artemis.utils.DataConstants;
+
+public class PageCountRecordV2 extends PageCountRecord {
```
Ditto as per other comment if v2 is needed.

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/QueuePendingMessageMetrics.java:
```diff
@@ -0,0 +1,108 @@
+
+public class QueuePendingMessageMetrics {
+
+   private final AtomicInteger messageCount = new AtomicInteger(0);
```
Could all these atomics be primitives with static atomic updates, just to reduce object size.

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/PageSubscription.java:
```diff
@@ -161,4 +165,9 @@
     */
    void onDeletePage(Page deletedPage) throws Exception;
+
+   AtomicLong getDeliveredCount();
```
not sure would want to expose an atomic outside the object as it allows other code to update it unexpectedly, should probably be  primitive and the object should return get on the atomic.

cshannon replied:
Good point, I will make that change.

cshannon commented:
@michaelandrepearce - I updated my PR, take a look and let me know what you think.

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/PagedReferenceImpl.java:
```diff
@@ -53,6 +53,8 @@
 
    private Object protocolData;
 
+   private Long messageSize;
```
Could this be primitive? and either -1 or 0 to denote empty or not set in logic, just objects = GC, and we're putting effort in to remove them on the hot path as much as possible to keep GC pressure as low as possible.

cshannon replied:
Yep, I'll set this to -1 to mean not set.

cshannon commented:
@michaelandrepearce - ok made those 2 changes so it is ready for another look

cshannon commented:
@clebertsuconic - No, I'm not trying to measure the size on the JVM.  I want to measure the size of the messages in general (including how much space is on disk) as a metric to know how much data is left to consume on the Queue.  The most accurate way is to use the encode size as this is the amount of space taken up on disk.  For large messages this also includes the body size.

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/codec/PageCountPendingImpl.java:
```diff
@@ -43,6 +44,8 @@ public PageCountPendingImpl(long queueID, long pageID, int inc) {
 
    long pageID;
 
+   long size;
```
Should default be -1

cshannon replied:
Why would we use -1 and not 0? These values default to 0 because we don't want the counter to change if the size is not set.  If we default to -1 then the counters will go negative where as if they are 0 when not set then nothing changes with the counters which is what we want.

michaelandrepearce replied:
There’s a difference between no size (0) and not being set (-1) the logic calculating should be aware of this as -1 means it would need to be calculated

michaelandrepearce replied:
Eg it’s typical for -1 to be used to represent null vs 0 as no size , slight semantic difference but important none the less

cshannon replied:
The issue is backwards compatibility.  These values are now being stored to the journal and old records won't have a size.  So by defaulting to a 0 value for when an old record is it makes everything simple.  Yes I could make it -1 but then I have to add checks for -1 versus just using 0 as a default and then the logic is clean as we don't need to handle negative edge cases.

cshannon replied:
@michaelandrepearce - did you have in mind having the getter in the class handle the -1 (ie have the getSize() method check for -1 and return 0 if negative) or having the getSize() method return the -1 as is and then have the code using the value explicitly do the check for negative?

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/codec/PageCountRecord.java:
```diff
@@ -26,18 +26,21 @@
 
    private long value;
 
+   private long size;
```
Default to -1

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/codec/PageCountRecordInc.java:
```diff
@@ -26,17 +26,20 @@
 
    private int value;
 
+   private long size;
```
Default to -1

clebertsuconic commented:
i would rename getSize() as getEncodeSize() on the interfaces then. getSize could be dubious and users may get the wrong expectation of the meaning.

cshannon commented:
@clebertsuconic - Fair enough, but I think I should call it getMessageSize() instead (i already did this in a couple places)  The reason why I would use that name is because the value is slightly different than getEncodeSize(). For the normal message case, getMessageSize() and geEncodeSize() are equal.  But for a large message getMessageSize() will be getEncodeSize() + the body size.

cshannon commented:
getPersistentSize() is fine with me as it specifies that the size represents the entire persisted size of the message

cshannon commented:
@clebertsuconic  - hmm so I'm starting to like you original idea of just using getEncodeSize() more as the the message might be non-durable and in that case the encode size is still used but it isn't on disk so that would be a little misleading.

However there is still the issue of what to do aboutt large messages?  Maybe we need a name like fullEncodeSize() or something so that it is implies it is the full message including the body?

clebertsuconic commented:
I would include the sum you mentioned for large message.  The large message file size plus the large message encode size.  That translates fine as getPersjstentsjze

cshannon commented:
Ok I can just use getPersistentSize() then for everything, I will update the PR shortly.

cshannon commented:
I updated the PR with getPersistentSize() everywhere along with comments to try and make it more clear what the value represented.  I can tweak it some more if people still think it's not clear.

michaelandrepearce commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/codec/PageCountRecordInc.java:
```diff
@@ -26,17 +26,20 @@
 
    private int value;
 
+   private long persistentSize;
```
Seems my comment is removed due to update. Should be -1.

Reason for this is that if it isn't set e.g. have read up an old journal as such decode wont set it, and you know you need to calculate it to get a correct value.

cshannon replied:
@michaelandrepearce  - That is fine I will make those values -1, but I don't think we would try and re-calculate a correct value.  When loading the journal you'd have to re-load the messages to calculate from the paging store which we don't want to do.  I think the proper behavior is to just ignore it and not use the value in the calculations.  Then over time as new records come in the values would be filled in and correct.  

If I change the value to -1 on the record, are you thinking we should just have the getter do the negative check and return 0 if it's -1 or were you thinking to just return -1 from the getter and then check for negative everywhere else the value is used?

clebertsuconic commented:
I think paging should be measured just in number of pages (files) on the address. That would give users an idea about storage consumption.  It wouldn’t indeed be practical to load it to calculate sizes.

cshannon commented:
@clebertsuconic - I don't think we need to re-calculate anything if the values are missing for paging.  The  point of this metric isn't to give the user information on storage consumption as that data already exists.  They can find out how many page files are being used and disk usage already, etc.

The point of these metrics is to know exactly how much data on a queue is left to consume.  So there might be a 5 megabyte page file but only 2 messages left to consume and the rest of the file is just waiting to be GC'd.  This metric only includes the 2 messages to be consumed and not the rest of the file.

clebertsuconic commented:
I was just saying paging metrics should be exposed. Even if the files are to be GCed but if something happened and they are not the user should have the metrics exposed about pages.  I think we do have them but i don’t remember now. I’m away from a computer and I can’t check now.

cshannon commented:
Oh ok I get it now.  Yeah you are right that the PagingStore interface has a method to get the number of pages and the page size which can be used to get an idea of consumption.  This value is exposed through AddressControl so users can see it.

clebertsuconic commented:
@cshannon although looking at the PR now you added persistent size into page counters... but as I read it, it is not persistent...


We should go all the way in.. i.e... persist the new size metric along with the page counters... or all the way back.. i.e... only give user's paging metrics through page size... 

Implementing the counter on the journal is no big deal.. but I would like to keep it compatible... older versions of the journal should still be compatible.

cshannon commented:
@clebertsuconic - The metric in the page counters is persisted along with the page counters now.  When the page counters are written to the journal that metric is encoded to disk and reloaded on restart.

 Older versions of the journal are compatible, they will just not have that size value loaded as it won't have been saved previously.

cshannon commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/persistence/impl/journal/codec/PageCountPendingImpl.java:
```diff
@@ -62,21 +67,41 @@ public long getPageID() {
    public void encode(ActiveMQBuffer buffer) {
       buffer.writeLong(queueID);
       buffer.writeLong(pageID);
+      buffer.writeLong(persistentSize);
```
@clebertsuconic - here is an example of where the size is now persisted

clebertsuconic replied:
OIC.. you already implemented it..


I would like to have a compatibility test validating that.. there's a journal test on the new compatibility testsuite already... but not one with paging.. and validating the counters...

If you have a way to do it.. it would be great.. otherwise let me know and I will do it.

cshannon replied:
I can add a compatibility test. The value will just be a default (@michaelandrepearce wants me to use -1 instead of 0 and i guess we just would ignore it if not set for the count).

Where's the journal test located in the test suite?

clebertsuconic replied:
JournalCompatibilityTest under compatibility-tests.


I don't think the test is using paging. we should just extend the test to use paging.


I can help you with that if you have questions.

cshannon commented:
@clebertsuconic - I did some cleanup work and I added a couple of basic tests to the compatibility test.  It doesn't do much but show that the current snapshot can load a 2.4 journal successfully (and it can load the paging store as well).  We can expand the tests with more if you want.

Also, I removed the persistentSize from the PageCountPendingImpl class as the size (and the count apparently) aren't actually used and not needed.  On recovery we can recover the size from the paged reference.  

Do you think this is ready to merge?

cshannon commented on artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/PageSubscriptionImpl.java:
```diff
@@ -439,7 +453,7 @@ public void confirmPosition(final Transaction tx, final PagePosition position) t
       confirmPosition(tx, reference.getPosition());
 
-      counter.increment(tx, -1);
+      counter.increment(tx, -1, -getPersistentSize(reference));
```
@michaelandrepearce  - So because of this line here we can't really default the persistentSize to -1, it needs to be 0.   This call here will store a new PageCountRecordInc() record to the journal with a negative count and negative size on message ack to decrement the counters. Because of this a negative can actually be a valid value and really 0 should mean not set (we should never have a message size of 0)

cshannon commented:
Had to update the PR again as something got messed up when rebasing
